### ISSUE TITLE
Unpacked conda environment breaks if moved

### ISSUE THREAD

**Issue by j-levy** [conda/conda-pack#106]
I noticed that the script conda-unpack only works once, which is problematic if the environment is moved to another location after being unpacked.
I did the following : 
- I pack my conda environment using conda-pack
- I untar the result in a given path, let's call it `/install/tar/path1/my_env`
- In this path, I run the `source ./bin/activate/`, then I run `conda-unpack`. At this point, conda-unpack rewrites the placeholders prefixes to match the path `/install/tar/path1`.
Everything works so far
- I move the folder `/install/tar/path1` to `/install/tar/path2`. 
- I go to `/install/tar/path2/my_env`, run `source ./bin/activate/`, then `conda-unpack`. But my environment doesn't work, binaries in my environment that are looking for their libs can't find them. I think this is because I already ran `conda-unpack` once, so the placeholders prefixes are already rewritten as `/install/tar/path1/my_env`, and now they don't match with `/install/tar/path2/my_env`.

Is there an option in conda (maybe in their file portability.py?) or in `conda-pack` to rewrite the local path so that `conda-unpack` always works in all cases? Or would you consider this as a potential future feature?

**Comment by j-levy** [comment]
currently working on this on my fork https://github.com/j-levy/conda-pack/tree/feature/repack-env
I added a --repack option to conda-unpack to reverse the path replacement. It works with my current project, but must be run before moving the environment folder.
I'll try to let the python script detect automatically whether the current path is the good one or not, and then do the replacement if needed.

**Comment by djhoese** [comment]
@j-levy Any updates on this? I started coding up a wrapper script for my project that would detect if the root path changed and then rerun conda-unpack. It looks like you are saying that this won't work which is too bad. Any idea what it would take to get a fix into conda-pack main.

**Comment by j-levy** [comment]
I did something that seems to work.

I log in a text file called current_path_patch.bak the path where the conda environment has been previously unpacked with conda-unpack. After moving the folder to a now location, running conda-unpack again will read the previous path in the text file and replace it in all files.
With text file, it's easy. With binary files, it doesn't cause any offsets problem thanks to the fact that there's always a lot of padding in the binary file (since that's how conda works, by setting the path in binary files with a lot of padding to be sure that relatively long paths can be replaced). Since we read the previous path in the text file, we know its length, so we can replace it.

(Originally I added an option --repack that updates the paths with the current directory when run but you can just run it directly with conda-unpack. If there's no current_path_patch.bak it will behave just like before.)

You can test it with the compiled version at https://anaconda.org/jolevy/conda-pack/
the code is here:  https://github.com/SED-Inria-Sophia/conda-pack/
Feel free to fork, adapt, clean up, and maybe file a Pull Request for it.
**(Note: this only works on Linux and macOS since conda-pack itself doesn't work on Windows at the moment, I don't know why.)**

**Comment by djhoese** [comment]
> since conda-pack itself doesn't work on Windows at the moment, I don't know why.)

It doesn't? I think I'm almost successfully using it on Windows. It packs it up, but I had some issues running the unpacked zip file. Wasn't sure it wasn't my fault though. Still debugging.

In my case I have one wrapper script (users are only expected to run one specific application from the environment) that does this unpack check and stores the current base path to a text file. I suppose even in my own version I could overwrite the unpack command to do this internally instead of in the wrapper script (and always run unpack when executing my wrapper). I'll take a look at your changes and see what I understand.

Have you made a PR for these changes yet to be merged upstream?

**Comment by mcg1969** [comment]
I'm skeptical that this is a wise inclusion into the tool. There are already multiple ways to "move" an existing environment: `conda clone` is likely a better fit for your application.

**Comment by djhoese** [comment]
You can do that on a unpacked conda environment?

Regardless, this functionality is needed for people who don't necessarily *know* that they are using a conda-pack'd environment and may not have any experience with conda at all. All they know is that they "installed" an application by unpacking the archive and can run it by executing a script/binary.

**Comment by mcg1969** [comment]
Conda environments should never be moved, ever. I suspect I'm preaching to the choir with you @djhoese but the original poster learned the hard way. It doesn't matter whether `conda` created the environment or `conda-pack` did, this just doesn't work. It's the very reason tools like `conda pack` and `conda create --clone` exist.

Fundamental to the design of both utilities is that they draw from the _original, unmodified package data_ when building the new environment or archive. Both utilities draw from the package cache when they do that. The reason for this is that the conda package contains specially prepared versions of the pre-relocated packages. This makes the moving/packaging process as robust as possible.

A key difference between the two options is that `conda create --clone` can re-download any missing entries into the package cache that it needs, while `conda pack` does not. In contrast, if the `conda pack` cannot find those package cache entries, it warns the user that it is dropping down into an "unmanaged" mode instead, which "usually" works. We shouldn't be in the business of "usually".

So for moving environments from one location on the same machine to another, `conda create --clone` is absolutely the right way to do it. It works both with environments originally created by `conda` as well as unpacked `conda-pack` archives. I don't think it would serve users well to offer a less robust version in `conda pack`.

**Comment by djhoese** [comment]
> Conda environments should never be moved, ever.

I get it. I do. It just feels so close to being much more flexible. I understand that this is likely a can of worms for maintenance/support of conda-pack though. The problem and confusion with this is that conda-pack is used for copying/relocating/distributing conda environments and comes with a utility (conda-unpack) to initialize a newly relocated environment. While maybe this is meant to only be used as a one-time thing, it is easy to think otherwise. Or it is easy for recipients of the conda-pack'd environments to not know the restrictions of the environment they are using. Again, I understand that conda-pack has no responsibility to make this work and support these use cases.

It's very confusing to have a tool for making "relocatable" environments that are not relocatable :stuck_out_tongue_winking_eye: Here's the first line of the website:

> conda-pack is a command line tool for creating relocatable conda environments.

I was surprised that the `conda create --clone` thing works when a conda-pack'd environment is specified since the conda-pack environment itself no longer has the conda utility (`bin/conda`). Cool. But also not available if this is the only conda-ish environment on a person's machine.

Bottom line, I get it. This isn't the intended purpose of conda-pack/unpack to allow people to unpack an environment and then move the directory around their machine. It seems like with some small modifications conda-unpack could do this...but ok. No biggie.

**Comment by mcg1969** [comment]
At the very least let's work on the docs to help eliminate the confusion. Clearly the result of conda-pack is an environment that can be unpacked anywhere, but that doesn't mean it is relocatable!

**Comment by mcg1969** [comment]
@djhoese I appreciate your understanding! It's important to understand that something that works "most of the time" for users still means a significant increase in support burden for maintainers for the times it _doesn't_ work. If someone offers a PR that is sufficiently generic and bullet proof, has good tests, etc. I think we should consider it.

In the meanwhile I absolutely agree we should clarify our language in the docs. I've created #130 to capture this issue and other doc issues. With that open I'll close this.
